internal/cluster/distribution/pke/pkeaws/pkeawsworkflow/activity_nlb_health_check.go

```diff
@@ -0,0 +1,143 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkeawsworkflow
+
+import (
+	"context"
+	"time"
+
+	"emperror.dev/errors"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"go.uber.org/cadence/activity"
+
+	"github.com/banzaicloud/pipeline/pkg/backoff"
+)
+
+const (
+	HealthCheckActivityName = "nlb-health-check"
+	backoffDelay            = 3 * time.Second
+	backoffMaxRetries       = 3
+)
+
+type HealthCheckActivity struct {
+	awsSessionFactory AWSSessionFactory
+	elbv2Factory      ELBV2APIFactory
+}
+
+// NewHealthCheckActivity returns a new HealthCheckActivity.
+func NewHealthCheckActivity(
+	awsSessionFactory AWSSessionFactory,
+	elbv2Factory ELBV2APIFactory,
+) *HealthCheckActivity {
+	return &HealthCheckActivity{
+		awsSessionFactory: awsSessionFactory,
+		elbv2Factory:      elbv2Factory,
+	}
+}
+
+type HealthCheckActivityInput struct {
+	OrganizationID uint
+	SecretID       string
+	Region         string
+	ClusterID      uint
+	ClusterName    string
+}
+
+// HealthCheckActivityOutput holds the output data
+type HealthCheckActivityOutput struct {
+	nlbStatus []*elbv2.TargetHealthDescription
+}
+
+// Register registers the activity in the worker.
+func (a *HealthCheckActivity) Register() {
+	activity.RegisterWithOptions(a.Execute, activity.RegisterOptions{Name: HealthCheckActivityName})
+}
+
+func (a *HealthCheckActivity) Execute(ctx context.Context, input HealthCheckActivityInput) (*HealthCheckActivityOutput, error) {
+	if a == nil {
+		return nil, errors.New("activity is nil")
+	}
+	session, err := a.awsSessionFactory.New(input.OrganizationID, input.SecretID, input.Region)
+	if err != nil {
+		return nil, errors.WrapIf(err, "failed to create AWS session")
+	}
+
+	elbv2Svc := a.elbv2Factory.New(session)
+	nlbName := "pke-" + input.ClusterName + "-nlb"
+
+	backoffConfig := backoff.ConstantBackoffConfig{
+		Delay:      backoffDelay,
+		MaxRetries: backoffMaxRetries,
+	}
+	backoffPolicy := backoff.NewConstantBackoffPolicy(backoffConfig)
+
+	var loadBalancerOutput *elbv2.DescribeLoadBalancersOutput
+	if err := backoff.Retry(func() error {
+		var errInt error
+		loadBalancerOutput, errInt = elbv2Svc.DescribeLoadBalancers(
+			&elbv2.DescribeLoadBalancersInput{
+				Names: []*string{&nlbName},
+			})
+		if errInt != nil {
+			return errors.WrapIf(errInt, "failed to describe load balancers")
+		}
+
+		return nil
+	}, backoffPolicy); err != nil {
+		return nil, err
+	}
+
+	var targetGroupOutput *elbv2.DescribeTargetGroupsOutput
+	if err := backoff.Retry(func() error {
+		var errInt error
+		targetGroupOutput, errInt = elbv2Svc.DescribeTargetGroups(
+			&elbv2.DescribeTargetGroupsInput{
+				LoadBalancerArn: loadBalancerOutput.LoadBalancers[0].LoadBalancerArn,
```

Comment:
**Required**: we **MUST** check the existence of the 0th value before trying to dereference it to avoid panic.

Same at
https://github.com/banzaicloud/pipeline/pull/3229/files#diff-07dfcca59ca7ddb4d7bf740160fa2bdc2bc1668420e7549b77e0106c15a728e1R123

Reply:
fixed: https://github.com/banzaicloud/pipeline/pull/3229/commits/db2f970c859fd030d62d292d475e7b86164f0478